From a72d8e19f0210dac94fec674cb9792beaf161b26 Mon Sep 17 00:00:00 2001
From: Greg Mann <greg@mesosphere.io>
Date: Wed, 1 Aug 2018 07:58:45 -0700
Subject: [PATCH 18/25] Enabled per-framework metrics in the allocator.

The `FrameworkMetrics` struct will hold per-framework metrics tracked
by the allocator.

Review: https://reviews.apache.org/r/66843/
---
 src/master/allocator/mesos/hierarchical.cpp |  3 ++-
 src/master/allocator/mesos/hierarchical.hpp |  2 ++
 src/master/allocator/mesos/metrics.cpp      |  7 +++++++
 src/master/allocator/mesos/metrics.hpp      | 10 ++++++++++
 4 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/master/allocator/mesos/hierarchical.cpp b/src/master/allocator/mesos/hierarchical.cpp
index dbfa08a01..bb36d406a 100644
--- a/src/master/allocator/mesos/hierarchical.cpp
+++ b/src/master/allocator/mesos/hierarchical.cpp
@@ -175,7 +175,8 @@ HierarchicalAllocatorProcess::Framework::Framework(
   : roles(protobuf::framework::getRoles(frameworkInfo)),
     suppressedRoles(_suppressedRoles),
     capabilities(frameworkInfo.capabilities()),
-    active(_active) {}
+    active(_active),
+    metrics(new FrameworkMetrics(frameworkInfo)) {}
 
 
 void HierarchicalAllocatorProcess::initialize(
diff --git a/src/master/allocator/mesos/hierarchical.hpp b/src/master/allocator/mesos/hierarchical.hpp
index 56d1d7d35..b4594199c 100644
--- a/src/master/allocator/mesos/hierarchical.hpp
+++ b/src/master/allocator/mesos/hierarchical.hpp
@@ -336,6 +336,8 @@ protected:
       inverseOfferFilters;
 
     bool active;
+
+    process::Owned<FrameworkMetrics> metrics;
   };
 
   double _event_queue_dispatches()
diff --git a/src/master/allocator/mesos/metrics.cpp b/src/master/allocator/mesos/metrics.cpp
index 5194f5b3b..7a4b153db 100644
--- a/src/master/allocator/mesos/metrics.cpp
+++ b/src/master/allocator/mesos/metrics.cpp
@@ -201,6 +201,13 @@ void Metrics::removeRole(const string& role)
   process::metrics::remove(gauge.get());
 }
 
+
+FrameworkMetrics::FrameworkMetrics(const FrameworkInfo& _frameworkInfo)
+  : frameworkInfo(_frameworkInfo) {}
+
+
+FrameworkMetrics::~FrameworkMetrics() {}
+
 } // namespace internal {
 } // namespace allocator {
 } // namespace master {
diff --git a/src/master/allocator/mesos/metrics.hpp b/src/master/allocator/mesos/metrics.hpp
index 6d386225c..daac93f7e 100644
--- a/src/master/allocator/mesos/metrics.hpp
+++ b/src/master/allocator/mesos/metrics.hpp
@@ -90,6 +90,16 @@ struct Metrics
   hashmap<std::string, process::metrics::PullGauge> offer_filters_active;
 };
 
+
+struct FrameworkMetrics
+{
+  explicit FrameworkMetrics(const FrameworkInfo& _frameworkInfo);
+
+  ~FrameworkMetrics();
+
+  const FrameworkInfo frameworkInfo;
+};
+
 } // namespace internal {
 } // namespace allocator {
 } // namespace master {
-- 
2.25.1

